lib/services/screen-builder-service.ts

```diff
@@ -77,34 +115,71 @@ export class ScreenBuilderService implements IScreenBuilderService {
+			let scaffolder = scaffolderData.scaffolder;
+			let type = screenBuilderOptions.type || ScreenBuilderService.DEFAULT_SCREENBUILDER_TYPE;
+			type = ScreenBuilderService.PREDEFINED_SCREENBUILDER_TYPES[type] || type;
+
+			if (type === ScreenBuilderService.DEFAULT_SCREENBUILDER_TYPE || screenBuilderOptions.answers) {
+					scaffolder.promptGenerate(type, screenBuilderOptions.answers, scaffolderData.callback);
+			} else {
+					scaffolder.promptGenerate(type, scaffolderData.callback);
+			}
 
-			let generators = this.$dependencyConfigService.getAllGenerators().wait();
-			_.each(generators, (generator: IGeneratorConfig) => this.$generatorExtensionsService.prepareGenerator(generator.name).wait());
+			return scaffolderData;
+		}).future<IScaffolder>()();
+	}
+
+	public ensureScreenBuilderProject(projectDir: string): IFuture<void> {
+		return (() => {
+			if(!_.every(ScreenBuilderService.SCREEN_BUILDER_SPECIFIC_FILES, file => this.$fs.exists(path.join(projectDir, file)).wait())) {
+				this.$errors.fail("This command is applicable only to Screen Builder projects.");
+			}
 		}).future<void>()();
 	}
 
-	private promptGenerate(generatorName: string, screenBuilderOptions?: IScreenBuilderOptions): IFuture<void> {
-		let scaffolderData = this.createScaffolder(generatorName, screenBuilderOptions).wait();
-		let scaffolder = scaffolderData.scaffolder;
-		let type = screenBuilderOptions.type || ScreenBuilderService.DEFAULT_SCREENBUILDER_TYPE;
-		type = ScreenBuilderService.PREDEFINED_SCREENBUILDER_TYPES[type] || type;
+	public shouldUpgrade(): IFuture<boolean> {
+		return (() => {
+			if (this.shouldUpgradeCached === null) {
+				let scaffolderData = this.createScaffolder(this.generatorName).wait();
+
+				scaffolderData.scaffolder.initContext({ collectMetadata: true }, scaffolderData.callback);
 
-		if (type === ScreenBuilderService.DEFAULT_SCREENBUILDER_TYPE || screenBuilderOptions.answers) {
-				scaffolder.promptGenerate(type, screenBuilderOptions.answers, scaffolderData.callback);
-		} else {
-				scaffolder.promptGenerate(type, scaffolderData.callback);
-		}
+				this.shouldUpgradeCached = scaffolderData.future.wait() === ScreenBuilderService.UPGRADE_ERROR_MESSAGE;
+			}
 
-		return scaffolderData.future;
+			return this.shouldUpgradeCached;
+		}).future<boolean>()();
 	}
 
-	private createScaffolder(generatorName: string, screenBuilderOptions?: IScreenBuilderOptions): IFuture<{ scaffolder: any; future: IFuture<any>; callback: Function }> {
+	public upgrade(): IFuture<void> {
+		return (() => {
+			if (!this.shouldUpgrade().wait()) {
+				return;
+			}
+
+			let scaffolderData = this.createScaffolder(this.generatorName).wait();
+
+			scaffolderData.scaffolder.upgrade(scaffolderData.callback);
+
+			scaffolderData.future.wait();
+		}).future<void>()();
+	}
+
+	private prepareScreenBuilder(): IFuture<void> {
+		return (() => {
+			this.$logger.out("Please, wait while Screen Builder and its dependencies are being configured.");
```

Comment:
Why do we have the same error message in two places? Is there a way to optimize this?